urlchecker/core/urlproc.py

```diff
@@ -90,28 +90,61 @@ def get_user_agent():
     return random.choice(agents)
 
 
+def find_urls(file_names, exclude_patterns=None, exclude_urls=None):
+    """
+    Given a list of file names, return the file name and a list of urls to parse
+    (intended to be passed on to the UrlCheckResultResult class) that removes
+    any redundant URLs.
+    """
+    # keep a master list of urls we have seen - no repeats allowed!
+    seen = set()
+
+    for file_name in file_names:
+        checker = UrlCheckResult(
+            file_name=file_name,
+            exclude_patterns=exclude_patterns,
+            exclude_urls=exclude_urls,
+        )
+
+        # Ensure we don't have repeats and updated seen list
+        checker.urls = [u for u in checker.urls if u not in seen]
+        [seen.add(u) for u in checker.urls]
+
+        # Skip empty sets of urls
+        if not checker.urls:
+            continue
+
+        # finder.urls is already unique
+        yield file_name, checker.urls
+
```

Comment:
I find the use of `yield` perfect. However, I haven't used it that often so I wonder if the `seen` variable will be further updated or reset-ed on each call. If so then maybe give `seen`to the function as an argument.

Reply:
> Overall this looks nice but can you maybe test it like you did with the previous PR so we can get a fair comparison. 

Sure! I don't expect to see differences here because the files we test 1. are pretty small in number, and 2. don't have huge redundant of URLs.

> I find the use of yield perfect. However, I haven't used it that often so I wonder if the seen variable will be further updated or reset-ed on each call. 

The seen variable remains in the function - the reason to have this design (with yield) is so we can update it between each yield.

Reply:
Example:
```python
def show_yield(): 
   ...:     seen = set() 
   ...:     for number in range(10): 
   ...:         yield number 
   ...:         print(seen) 
   ...:         seen.add(number) 
```
```python
In [4]: for number in show_yield(): 
   ...:     continue 
   ...:                                                                                                                                      
set()
{0}
{0, 1}
{0, 1, 2}
{0, 1, 2, 3}
{0, 1, 2, 3, 4}
{0, 1, 2, 3, 4, 5}
{0, 1, 2, 3, 4, 5, 6}
{0, 1, 2, 3, 4, 5, 6, 7}
{0, 1, 2, 3, 4, 5, 6, 7, 8}
```

Reply:
looks good to me :))

Reply:
okay sounds good! I'll build an action container for this tomorrow (err in a few hours, still~5am here) and give it a test run on USRSE. We definitely might see variance so I'm not sure how much improvement we can measure, but it's worth a try!